proxylib/redis/redisparser.go

```diff
@@ -0,0 +1,447 @@
+			}
+		}
+
+		// ensure that 'cmd' value is a valid cmd and that key is
+		// only provided in scenarios when it makes sense.
+
+		if len(rr.cmdExact) > 0 {
+			// ensure this is a valid cmd
+			res := cmdMap[rr.cmdExact]
+			if res == cmdInvalid {
+				ParseError(fmt.Sprintf("Unable to parse L7 redis rule with invalid cmd: '%s'",
+					rr.cmdExact), rule)
+			} else if res == cmdNoKey && rr.keyRegexCompiled != nil {
+				ParseError(fmt.Sprintf("key regex not allowed for rule with cmd '%s'", rr.cmdExact), rule)
+			} else if res == cmdSpecialKey && rr.keyRegexCompiled != nil {
+				// this is temporary, as in the future we will properly parse a list of keys
+				// from these commands, and potentially allow a list of keys to match
+				ParseError(fmt.Sprintf("key regex not allowed for rule with cmd '%s'", rr.cmdExact), rule)
+			}
+
+		}
+
+		log.Debugf("Parsed RedisRule pair: %v", rr)
+		rules = append(rules, &rr)
+	}
+	return rules
+}
+
+type RedisParserFactory struct{}
+
+var redisParserFactory *RedisParserFactory
+
+func init() {
+	log.Info("init(): Registering redisParserFactory")
+	RegisterParserFactory("redis", redisParserFactory)
+	RegisterL7RuleParser("redis", RedisRuleParser)
+}
+
+type RedisParser struct {
+	connection *Connection
+	inserted   bool
+}
+
+/*
+type RedisRequestInfo struct {
+	cmd string
+	keys []string
+}
+*/
+
+const accessDeniedStr = "-Error Access Denied\r\n"
+
+func (pf *RedisParserFactory) Create(connection *Connection) Parser {
+	log.Debugf("RedisParserFactory: Create: %v", connection)
+
+	p := RedisParser{connection: connection}
+	return &p
+}
+
+func (p *RedisParser) OnData(reply, endStream bool, dataArray [][]byte) (OpType, int) {
+
+	// inefficient, but simple for now
+	data := bytes.Join(dataArray, []byte{})
+	if len(data) == 0 {
+		return MORE, 1
+	}
+
+	if reply {
+		// We currently only look at the request stream, not the reply
+		// stream.  So technically we could "PASS" with a much higher number
+		// here to improve efficiency.  At least at this point, we will keep
+		// pass the data here for visibility.
+		log.Infof("OnData: <== '%s'", data)
+		return PASS, len(data)
+	}
+
+	log.Debugf("OnData: ==> '%s'", data)
+
+	var consumed int
+	cmd, err := readSingleCommand(data, &consumed)
+	if err != nil {
+		log.Infof("Parse error: %v", err)
+		return ERROR, int(ERROR_INVALID_FRAME_LENGTH)
+	}
+	log.Infof("consumed = %d", consumed)
+	if consumed == 0 {
+		return MORE, 1
+	}
+	// if consumed is greater than zero, then we should
+	// have at least one command
+	if len(cmd.Args) == 0 {
+		log.Infof("Consumed %d bytes, but have zero length cmd array", consumed)
+		return ERROR, int(ERROR_INVALID_FRAME_LENGTH)
+	}
+	log.Infof("CMD Args = %v", cmd.Args)
+
+	access_log_entry_type := cilium.EntryType_Request
+	parser_action := PASS
+
+	if !p.connection.Matches(cmd) {
+		access_log_entry_type = cilium.EntryType_Denied
+		parser_action = DROP
+		p.connection.Inject(true, []byte(accessDeniedStr))
```

Comment:
It seems it is possible for the injected error reply to be placed ahead of a response pending from the server. Does it matter for redis?

Reply:
@jrajahalme good point.   Redis is typically request/response, but it does support pipelining (https://redis.io/topics/pipelining), in which case we would have to deal with this.